cql3/expr/expression.hh

```diff
@@ -449,6 +449,9 @@ using value_set = std::variant<value_list, nonwrapping_range<managed_bytes>>;
 /// - `A=1 AND A<=0` restricts A to an empty list; no value is able to satisfy the expression
 /// - `A>=NULL` also restricts A to an empty list; all comparisons to NULL are false
 /// - an expression without A "restricts" A to unbounded range
+///
+/// In case of a subscripted value, e.g `mycolumn[5] = 6` you should pass `mycolumn` as the first argument
+/// to get possible lhs values.
```

Comment:
When does this happen? For an index on a constant subscript?

Reply:
Sorry I must've missed your comment.

I was sure that I saw this happening once in a test case that had a frozen map column in a partition key.
It went something like:

```cql
cqlsh:ks> CREATE TABLE t (m frozen<map<int, int>> primary key);
cqlsh:ks> SELECT * FROM t WHERE m[0] = 0 ALLOW FILTERING;
```
And then `possible_lhs_values` was called inside this function: https://github.com/scylladb/scylla/blob/c89117e5f75846c1ee5c236f424260608b3eda6a/cql3/restrictions/statement_restrictions.cc#L822

But now I see that trying to do something like that throws an error:
```
InvalidRequest: Error from server: code=2200 [Invalid query] message="Map-entry equality predicates on frozen map column m are not supported"
```

I vividly remember fighting a bug where this function threw `bad_variant_access` because it expected `column_value` instead of `subscript`, but I guess my memory must be faulty.

Looking at it again it looks like `possible_lhs_values` is never actually called with a `subscript` (at least not in `dev-test`).
I guess we could throw an error instead of trying to handle it, but retaining this behaviour would be much safer - we can be sure that nothing breaks in some test case that we didn't think of.

I think we should go ahead with this PR, even if it looks like `possible_lhs_values` is never called with  `subscript`.
The behaviour might be a bit strange, but it isn't too bad.

Reply:
Well, I don't understand it.

possible_lhs_values is used to jump out of the filtering path (where the expression is evaluated for every row) and into a better path (by computing the set of values that would make the expression true, and using it as a partition key / clustering key range / index key). But it doesn't fit in this case.

Reply:
Yeah it is a bit inelegant but safe. Basically we have three options:

* **A**
  We keep the old behaviour: `possible_lhs_values('col', "col[2] = 3")` returns `{3}`.
  This is a bit ugly - we ask for values for `col`, but instead we get values for `col[2]`.
  The advantage is that all existing code will work and it isn't really THAT ugly.
  
* **B**
  We disallow/ignore `subscript` as argument to `possible_lhs_values`.
  This would be more correct, but the problem is that the existing code might break in some unexpected case that is not covered by `dev-test`.
  It's easy to imagine that there might be some code which calls `possible_lhs_values` without caring that something might be a subscript - after all this used to work without a problem. It might be hard to ensure that this doesn't make scylla crash on some random users query.

* **C**
  We refactor `possible_lhs_values` to handle `subscripts` more cleanly.
  This could be a good idea, but for now I feel like this could end up as trying to predict the future.
  Maybe it would be better to do it while implementing indexes on collections. The person doing that would know exactly what they need from this function and how to refactor it to meet their needs.

I would go with `A`

Reply:
I prefer B (with on_internal_error). Leaving it like that is a time bomb.

For C, I think the refactoring should transform the original restriction into a new expression that evaluates to the lhs values, rather than to the values themselves. It's not directly possible because possible_lhs_values returns a value_set, not a single value.

Some examples:

1. WHERE a = :a    -> :a

Simplest case

2. WHERE a = :a AND non_pk_column = :b  -> :a

non_pk_column doesn't affect the result so we can remove it

3. WHERE a = :a AND b = :b -> tuple(:a, :b)

Instead of making each pk column compute possible_lhs_values() independently, have possible_lhs_values compute the entire pk as one expression

4. WHERE a IN :a_list AND b IN :b_list -> cartesian_product(:a_list, :b_list)

We'd transform the where clause into something like

struct prepared_where {
    expression pk_expression; -> goes into partition_range_vector
    bool pk_is_list; // if pk_expression returns a list of tuples rather than tuples
    expression ck_expression; // returns list(tuple(lower_bound, higher_bound)), goes into partition_slice
    expression filter; // everything that remains, evaluated for each row
}